test/cppsuite/tests/hs_cleanup.cpp

```diff
@@ -52,44 +53,35 @@ class hs_cleanup : public test {
         logger::log_msg(
           LOG_INFO, type_string(tc->type) + " thread {" + std::to_string(tc->id) + "} commencing.");
 
-        const char *key_tmp;
         const uint64_t MAX_ROLLBACKS = 100;
         uint32_t rollback_retries = 0;
 
-        collection &coll = tc->db.get_collection(tc->id);
-
         /* In this test each thread gets a single collection. */
         testutil_assert(tc->db.get_collection_count() == tc->thread_count);
+
+        collection &coll = tc->db.get_collection(tc->id);
         scoped_cursor cursor = tc->session.open_scoped_cursor(coll.name);
 
-        /* We don't know the keyrange we're operating over here so we can't be much smarter here. */
+        /*
+         * We don't know the key range we're operating over here so we can't be much smarter here.
+         */
         while (tc->running()) {
             tc->sleep();
 
+            /* Start a transaction if possible. */
+            tc->txn.try_begin();
+
             auto ret = cursor->next(cursor.get());
             if (ret != 0) {
-                if (ret == WT_NOTFOUND) {
-                    testutil_check(cursor->reset(cursor.get()));
-                    continue;
-                }
-                if (ret == WT_ROLLBACK) {
-                    /*
-                     * As a result of the logic in this test its possible that the previous next
-                     * call can happen outside the context of a transaction. Assert that we are in
-                     * one if we got a rollback.
-                     */
-                    testutil_check(tc->txn.can_rollback());
-                    tc->txn.rollback();
-                    continue;
-                }
-                testutil_die(ret, "Unexpected error returned from cursor->next()");
+                testutil_assert(ret == WT_NOTFOUND || ret == WT_ROLLBACK);
+                testutil_check(cursor->reset(cursor.get()));
+                tc->txn.rollback();
```

Comment:
It's a bit confusing that we rollback on a `WT_NOTFOUND` here. Is this necessary?

Reply:
Ultimately, what I want is to restart the txn if I get an error that's why I trigger the rollback code.
@luke-pearson, do you think it's a misuse of the API?

Reply:
If you rollback here you are potentially ignoring the read_operation op_count. Which is supposed to be `ops_per_transaction=(max=1000,min=1),`, this isn't a misuse of the API but it is a misuse of the framework.

Reply:
The `read_operation` should not be considered here since this is the `update_operation` code. However, I agree we should not roll back if we get `WT_NOTFOUND` as we may have performed updates before getting this error code, it's better to `reset` the cursor instead.
See 96fde86d584745833f50c00bb2e2c5b7babbd20b